### ISSUE TITLE
omc-based FMI import in OMEdit

### ISSUE THREAD

**Issue by OpenModelica-TracImporter** [OpenModelica/OpenModelica#5345]
In today’s developer meeting we discussed to removing the omc-based FMI import.

As a first step, we agreed on replacing the omc-based FMI import in OMEdit with a pointer to the OMSimulator documentation.

If there is a good reason to keep the omc-based fmi import, please let us know. Otherwise the functionality will be completely replaced by OMSimulator at a later point in time.

------------------------

From https://trac.openmodelica.org/OpenModelica/ticket/5345

**Comment by casella** [comment]
Summary of previous discussion:
- enabling full-fledged FMI import into Modelica models in OMEdit would require extra work, and may actually not be technically possible in some cases
- FMI import can nevertheless be very useful in many cases that are indeed supported by the current implementation
- there is definitely an end-user requirement for that, see, e.g., the discussion in  [#5621](https://trac.openmodelica.org/OpenModelica/ticket/5621#comment:26) and [#5345](https://trac.openmodelica.org/OpenModelica/ticket/5345)
- on the other hand, FMI import currently fails in many cases, leaving the end users puzzled
- full-fledged, no compromise simulation of FMUs is only possible within an FMI/SSP tool such as OMSimulator

Proposal:
- we keep the current implementation
- we figure out exactly what cases are not covered
- we write them down in the documentation
- if possible, we make sure that OMEdit warns the user if the applicability conditions are not met

@casella can take care of the documentation part. @lochel, @adrpo, please add a comment to this ticket with a list of FMI features or particular cases where FMI import doesn't work.

**Comment by casella** [comment]
First of all, this feature is currently broken, so until it is fixed it should be removed from the documentation, see #9443.

I re-read again the discussion in [#5621](https://trac.openmodelica.org/OpenModelica/ticket/5621) and in #7591. I'll try to summarize the main takeaways in a single place here, because it is quite a long story.

1. FMI import works through a wrapper that calls `reinit()` in an algorithm. This was accepted by the old frontend and worked successfully in several applications, but is no longer accepted by the new frontend, because it is not legal Modelica. Therefore, as the old frontend is retired for the purpose of compiling simulation models and no longer maintained, this feature is currently unsupported.
2. This feature can be useful to integrate FMI blocks inside hierarchically structured Modelica models and was used successfully with older version of OMEdit, so it is obviously a problem if it is no longer supported. This is argued by @max-privato (a.k.a @ceraolo) e.g. [here](https://trac.openmodelica.org/OpenModelica/ticket/5621#comment:18), [here](https://trac.openmodelica.org/OpenModelica/ticket/5621#comment:20), and [here](https://trac.openmodelica.org/OpenModelica/ticket/5621#comment:27).
3. @lochel claimed multiple times that the proper way to solve this issue is using OMSimulator, e.g. [here](https://trac.openmodelica.org/OpenModelica/ticket/5621#comment:19), [here](https://github.com/OpenModelica/OpenModelica/issues/7591#issuecomment-865658655), and [here](https://github.com/OpenModelica/OpenModelica/issues/7591#issuecomment-865885829). Unfortunately @lochel stopped working directly for OSMC this year before the meeting to clarify this design was held, and it seems to me that the solution he envisions (which is still not too clear to me) would anyway requires quite some work on the GUI to be implemented, and we currently do not have the resources for that. Maybe we should try to hold that meeting ASAP anyway to clarify the design and estimate the effort 😃.
4. The issue of FMU import into Modelica is obviously an important goal for the Modelica Association (see modelica/ModelicaSpecification#1727 and modelica/ModelicaSpecification#2389) and should be preferably resolved at that level. Unfortunately, the interest on such a topic seems to be quite low, so this is not going to happen anytime soon.
5. In general, importing an FMU into Modelica may be impossibile, because the model representation of FMUs (also ME-FMUs) is a computational one, not a declarative one, allowing for concepts like discarding one step and going back, which has no equivlent in Modelica. That's what makes the goal of modelica/ModelicaSpecification#1727 difficult. However, there is certainly a sub-set of FMUs (I guess all continuous-time blocks, for example) that are indeed fully equivalent to a Modelica block. The models mentioned by @max-privato and referenced in point 2. are proof of this, since they used to work fine with old versions of OMC. For them, having an FMU-import facility in newer versions of OMC than 1.14.1 would be quite useful.
6. One possible solution would be to try to remove the `reinit()` statement from our wrapper implementation, as suggested [here](https://github.com/OpenModelica/OpenModelica/issues/7591#issuecomment-865866506), so that it does not contain non-standard Modelica code. However, it is not clear if this is even possible, who could do that, and how much time it would take.
7. Another solution (endorsed by @adrpo [here](https://github.com/OpenModelica/OpenModelica/issues/7591#issuecomment-865487312)) is to add an exception in omc to handle the `reinit()` statement in the current wrapper, thus re-enabling the formerly available solution. According to #perost's [comment](https://github.com/OpenModelica/OpenModelica/issues/7591#issuecomment-865738573) this requires some work, but he estimates [a couple of hours](https://github.com/OpenModelica/OpenModelica/issues/7591#issuecomment-865753384), so it's no big deal. @sjoelund rightly notices [here](https://github.com/OpenModelica/OpenModelica/issues/7591#issuecomment-865658872) that this requires non-standard Modelica, so it is possibly non-portable, and it should be handled consistently with the way we handle non-standard constructs, by means of the [--allowNonStandardModelica](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-allownonstandardmodelica) flag.

Even though I agree that solution 7. is not perfect, because it is not portable (due to the use of non-standard Modelica in the wrapper) and not fully general (not all FMUs will run), it can indeed be useful in many use cases, and it's the only one that can realistically be implemented _with a very limited effort_. So, I see no reason not to go for it.

So, here is my proposal
- [ ] @perost can implement the support for `reinit()` in algorithm sections, activated with the `--allowNonStandardModelica=reinitInAlgorithms` flag. The error message issued if this flag is not activated should suggest to do so, and a warning about this not being standard be issued when the flag is activated
- [ ] @adeas31 can add one checkbox to Translation Flags "Enable FMU Import"
- [ ] @casella can update the `documentation`, explaining that this feature only works with some FMUs (e.g. continuous time blocks), is not guaranteed to work in general, and is not portable, because it uses non-standard Modelica code in the FMU wrapper

@perost, @adeas31 can we fit this in 1.20.0? Otherwise we need to proceed with #9443 before we release 1.20.0.

If and when a better solution based on OMSimulator becomes available, we could then deprecate this solution and suggest to use the better one.

**Comment by perost** [comment]
>     * [x]  @perost can implement the support for `reinit()` in algorithm sections, activated with the `--allowNonStandardModelica=reinitInAlgorithms` flag. The error message issued if this flag is not activated should suggest to do so, and a warning about this not being standard be issued when the flag is activated

Done in #9525, except that I missed that you wanted the error message to be changed. Adding a warning about it not being standard is unnecessary I think though, since the user has to use a specific flag to enable it.

**Comment by perost** [comment]
#9528 changes the error message for reinit statements to mention the flag.

**Comment by perost** [comment]
> The `reinit` error doesn't show up with the flag `--allowNonStandardModelica=reinitInAlgorithms` but now the simulation fails with error,
> 
> `module = BouncingBall, log level = ERROR: [logStatusError][FMU status:Error] fmi2SetTime: Illegal model exchange call sequence. fmi2SetTime is not allowed in model_state_instantiated state.`

Does it work using the old frontend?

**Comment by casella** [comment]
Ouch. I hope we weren't also using non-standard FMI API calls in the wrapper, besides non-standard Modelica calls, that were previously not caught as illegal by the FMI code we exported. Then maybe @lochel improved FMI export by making it more compliant to the FMI spec, and we didn't catch the fact that it broke the FMI import wrapper because this was masked by the NF issue.

**Comment by casella** [comment]
BTW, the current example of FMI import in the documentation  is quite misleading:
- it imports a stand-alone system, which is best simulated using OMSimulator
- the FMU has a reinit statement, which is problematic

I guess it may be better to export an FMI with some inputs and outputs and only continuous-time dynamics, and then demonstrate it in a Modelica model where it is imported and connected to something else.

I will try out something tomorrow and see if it can work well, in which case we can document that instead.

**Comment by casella** [comment]
> Does it work using the old frontend?

Apparently not, alas. I tried a simple example myself, exporting a Modelica.Blocks PI controller and then reimporting it in a simple feedback control system. Only continuous-time dynamics. I get exactly the same error:
```
fmi2SetTime failed with status : Error

Process crashed
module = TestFMIImport_PIController, log level = ERROR: [logStatusError][FMU status:Error] fmi2SetTime: Illegal model exchange call sequence. fmi2SetTime is not allowed in model_state_instantiated state.
```

**Comment by casella** [comment]
I think I figured out what is still going wrong here, though I'm not much into FMI details myself.

Our Modelica FMU wrapper starts with these initial equations:
```modelica
  flowStartTime = fmi2Functions.fmi2SetTime(fmi2me, time, 1);
  flowEnterInitialization = fmi2Functions.fmi2EnterInitialization(fmi2me, flowParamsStart+flowInitInputs+flowStartTime);
  flowInitialized = fmi2Functions.fmi2ExitInitialization(fmi2me, flowParamsStart+flowInitInputs+flowStartTime+flowEnterInitialization);
  fmi_x = fmi2Functions.fmi2GetContinuousStates(fmi2me, numberOfContinuousStates, flowParamsStart+flowInitialized);
```
I understand the summed inputs are meant to make sure that dependency analysis forces the calling of the RHS functions with a certain well-defined order. In this case, the sorting will be such that `fmi2SetTime()` will be called _before_ `fmi2EnterInitialization()`.

If I simply replace the first initial equation with
```
  flowStartTime = 0;
```
the simulation of my closed-loop test case with imported FMI PI controller works fine, and produces exactly the same results as the ones I obtain with a pure Modelica implementation. Hence, the problem lies exactly in the function call of that first initial equation. 

This is the implementation of [fmi2SetTime()](https://github.com/OpenModelica/OpenModelica/blob/12071b12048a3aca19e8fdda12552b97b8a6d455/OMCompiler/SimulationRuntime/fmi/export/openmodelica/fmu2_model_interface.c.inc#L1681)
```C
fmi2Status fmi2SetTime(fmi2Component c, fmi2Real t)
{
  ModelInstance *comp = (ModelInstance *)c;
  if (invalidState(comp, "fmi2SetTime", model_state_me_event_mode|model_state_me_continuous_time_mode, 0))
    return fmi2Error;
  FILTERED_LOG(comp, fmi2OK, LOG_FMI2_CALL, "fmi2SetTime: time=%.16g", t)
  comp->fmuData->localData[0]->timeValue = t;
  comp->_need_update = 1;
  return fmi2OK;
}
```
This is consistent with the table at page 89 of FMI 2.0.2, which shows that `fmi2SetTime()` can indeed be called only in those two states. From the state machine at page 87, from the table at page 89, and from the table at page 102, I understand that the initial time should instead by set by calling `fmi2SetupExperiment()`. 

This is also confirmed at page 23, where it is stated that
> `fmi2SetupExperiment` must be called at least once before calling `fmi2EnterInitializationMode` , in order that `startTime` is defined

So, as I understand, the current wrapper was abusing `fmi2SetTime()`, which was apparently allowed by our previous implementation of FMU export code. @lochel corrected that (among other things) in 80312003, that removed `model_instantiated` from the set of valid states from which you can call `fmi2SetTime()`. This happened 7 months ago, when the default NF had already broken the FMI-import feature for a long time, so we didn't notice.

This is good, because of course we want to be able to import FMUs from other tools (that's the whole point), that will also not allow calling `fmi2SetTime()` illegally.

Bottom line: in order to get FMI import back to life, we should also
- add an external function to the FMU import wrapper, to call `fmi2SetupExperiment()`
- call that function instead of fmi2SetTime in the first initial equation of the wrapper to set the initial time. I guess all other inputs to that function are not important, so we could give them some default value.

@adeas31 I see you were involved in these chunks of code some time ago. Could you take care of doing that, possibly with help from @adrpo and @arun3688? 

I can provide some simple test cases to make sure that simple exported and re-imported blocks still work as expected. The one I mentioned above is this one: [TestFMIImport.zip](https://github.com/OpenModelica/OpenModelica/files/9805205/TestFMIImport.zip).

**Comment by ceraolo** [comment]
I've checked today on  v1.20.0-dev-394 and found that you've done a great job!

I've made a few tests, with the following procedure:
1. consider a model containing a casual subsystem and run it
2. create a FMU from the casual system
3. import the FMU, insert it into a copy of the original model, and simulate again
4. check the compliance of the results with those obtained in step 1.
5. change parameters rather randomly in the original model and the one with the FMU and verify that the compliance is maintained.

The results were perfectly compliant in both steps 4 and 5.

---
I have two minor improvements to propose, though. 

1) when we use a model containing an imported FMU we the the following message:

`Operator reinit may not be used in an algorithm section (use flag --allowNonStandardModelica=reinitInAlgorithms to ignore).
`
this is good enough; but since OMEdit distinguishes between Translation Flags and Simulation Flags, why not change the message to the following?

`Operator reinit may not be used in an algorithm section (use translation flag --allowNonStandardModelica=reinitInAlgorithms to ignore).
`

2) Adding `--allowNonStandardModelica=reinitInAlgorithms` among Translation Flags in OMEdit works good. But even though we click on "Save experiment annotation inside model", this flag is not saved, maybe because there not exist such flag among the __Openmodelica custom flags. 
What about adding this flag to the list? This would ease operations when starting a new session.

**Comment by adeas31** [comment]
> 2. Adding --allowNonStandardModelica=reinitInAlgorithms among Translation Flags in OMEdit works good. But even though we click on "Save experiment annotation inside model", this flag is not saved, maybe because there not exist such flag among the __Openmodelica custom flags.
What about adding this flag to the list? This would ease operations when starting a new session.

Its part of `Save translation flags inside model i.e., __OpenModelica_commandLineOptions annotation` and not `Save experiment annotation inside model`.

**Comment by ceraolo** [comment]
aha, my fault, thanks.

**Comment by casella** [comment]
@ceraolo I improved the error message as you suggested in #9602.

**Comment by ceraolo** [comment]
I checked again. I also saved the option `--allowNonStandardModelica=reinitInAlgorithms` on the model successfully through OMEdit GUI.
Everthing looks great!

**Comment by casella** [comment]
We still have issues with fmi2SetTime, see #9918.

**Comment by casella** [comment]
@ceraolo reports in #9918 that he still gets errors from Dymola-generated FMUs due to a not allowed `fmi2SetTime` call.

In #9524, we changed the verboten call
```modelica
  initial equation
    flowStartTime = fmi2Functions.fmi2SetTime(fmi2me, time, 1);
```
with the kosher call
```modelica
  initial equation
    flowStartTime =  fmi2Functions.fmi2SetupExperiment(fmi2me, false, 0.0, time, false, 0.0, flowParamsStart+flowInitInputs);
```
However, there is still a call to `fmi2SetTime` in the [algorithm section](https://github.com/OpenModelica/OpenModelica/blob/917eecbd90f8244ef271dd06f3a3039cd975be34/OMCompiler/Compiler/Template/CodegenFMU.tpl#L2137):
```modelica
  algorithm
    flowTime := fmi2Functions.fmi2SetTime(fmi2me, time, flowInitialized);
```
This is OK during simulation, but we forgot that when the initial equations are solved, they also include whatever is found in the `equation` and `algorithm` sections, not only in the `initial equation` and `initial algorithm` sections 😅. Hence, `fmi2SetTime` is called at least once during initialization, which is verboten.

So, I guess [line 2137 of CodegenFMU.tpl](https://github.com/OpenModelica/OpenModelica/blob/917eecbd90f8244ef271dd06f3a3039cd975be34/OMCompiler/Compiler/Template/CodegenFMU.tpl#L2137) should be something like:
```
  algorithm
    flowTime := if not initial() then fmi2Functions.fmi2SetTime(fmi2me, time, flowInitialized) else time;
```
so that this function is only actually called during simulation, not during initialization, where the call to fmi2SetupExperiment should be enough to set the time value properly.

I opened PR #9934 for that. @adeas31, @arun3688 please check before it's merged.

@ceraolo, once PR #9934 is merged in, you can try with the next nightly build. Please make sure you build the test model again from scratch, because the one you currently have was created with a wrong internal Modelica code, which won't be affected by this fix - only FMUs imported anew will. 

@ceraolo could then test this in the nightly build, and if it works we put it in maintenance/v1.20, so we get it in 1.20.1.

**Comment by ceraolo** [comment]
Did what @adeas31 proposes, but got error (still 1.20.0):
![image](https://user-images.githubusercontent.com/7373030/207288668-e3c49cdf-f89e-4615-b972-91efb40d1d2d.png)

**Comment by ceraolo** [comment]
I also repeated the original test from ticket #9918 with 1.21.0-dev-148: still got the same error.

**Comment by adeas31** [comment]
Can you share the generated code?

**Comment by casella** [comment]
This is weird, as far as I understand, there should be no other calls to `fmi2SetTime` there. @ceraolo please post the generated Modelica code so we can double-check it.

**Comment by ceraolo** [comment]
[FirstDym.zip](https://github.com/OpenModelica/OpenModelica/files/10218172/FirstDym.zip)

**Comment by casella** [comment]
@ceraolo, I meant the Modelica code, not the FMU 😅

**Comment by ceraolo** [comment]
I added Modelica code, fmus, and imported fmus,  to the original ticket #9918. I repeat it here, where I added a further file for more clarity:
[FirstDym.zip](https://github.com/OpenModelica/OpenModelica/files/10225582/FirstDym.zip)

**Comment by ceraolo** [comment]
I added a zip  to the original ticket #9918 with several files.
I don't know exactly what you needed more, so now I put in the zip all possible files:
- First.fmu and FyrstDym.fmu: the two fmus
- First.mo and StepFirst.mo: modelica files without automatic generated modelica from imports
-  FirstFmu and FirstFmuDym: the modelica files generated when importing fmus
- StepFirstFmu and StepFirstFmuDym: the files using the imported fmus
[FirstDym.zip](https://github.com/OpenModelica/OpenModelica/files/10225717/FirstDym.zip)

**Comment by ceraolo** [comment]
> I added a zip to the original ticket #9918 with several files. I don't know exactly what you needed more, so now I put in the zip all possible files:
> 
> * First.fmu and FyrstDym.fmu: the two fmus
> * First.mo and StepFirst.mo: modelica files without automatic generated modelica from imports
> * FirstFmu and FirstFmuDym: the modelica files generated when importing fmus
> * StepFirstFmu and StepFirstFmuDym: the files using the imported fmus
>   [FirstDym.zip](https://github.com/OpenModelica/OpenModelica/files/10225717/FirstDym.zip)

I checked on today's nightly (dev-164), but received the same error message as in    https://github.com/OpenModelica/OpenModelica/issues/5345#issuecomment-1348110890.

**Comment by sergiochavezlazo96** [comment]
> Did what @adeas31 proposes, but got error (still 1.20.0): ![image](https://user-images.githubusercontent.com/7373030/207288668-e3c49cdf-f89e-4615-b972-91efb40d1d2d.png)

This is a similar to the error I am getting in [this ](https://github.com/OpenModelica/OpenModelica/issues/12584) issue. In my case I am try to run embedded python code and it crash only with non-standard libraries (even though it is able to load them)
